08-IO/Party.hs

```diff
@@ -0,0 +1,181 @@
+assignment.)
+-}
+--Exercise 2
+treeFold :: (a -> [b] -> b) -> b -> Tree a -> b
+treeFold f init (Node val children) = f val (map (treeFold f init) children)
+{-
+The algorithm
+Now let’s actually derive an algorithm to solve this problem. Clearly
+there must be some sort of recursion involved—in fact, it seems that
+we should be able to do it with a fold. This makes sense though— 
+starting from the bottom of the tree and working our way up, we
+compute the best guest list for each subtree and somehow combine
+these to decide on the guest list for the next level up, and so on. So
+we need to write a combining function
+combineGLs :: Employee -> [GuestList] -> GuestList
+which takes an employee (the boss of some division) and the optimal
+guest list for each subdivision under him, and somehow combines
+this information to compute the best guest list for the entire division.
+However, this obvious first attempt fails! The problem is that we
+don’t get enough information from the recursive calls. If the best
+guest list for some subtree involves inviting that subtree’s boss, then
+we are stuck, since we might want to consider inviting the boss of the
+entire tree—in which case we don’t want to invite any of the subtree
+bosses (since they wouldn’t have any fun anyway). But we might be
+able to do better than just taking the best possible guest list for each
+subtree and then excluding their bosses.
+The solution is to generalize the recursion to compute more information, in such a way that we can actually make the recursive step.
+In particular, instead of just computing the best guest list for a given
+tree, we will compute two guest lists:
+1. the best possible guest list we can create if we invite the boss (that
+is, the Employee at the root of the tree); and
+2. the best possible guest list we can create if we don’t invite the boss.
+It turns out that this gives us enough information at each step to
+compute the optimal two guest lists for the next level up.
+-}
+
+{-
+Exercise 3
+Write a function
+nextLevel :: Employee -> [(GuestList, GuestList)]
+-> (GuestList, GuestList)
+which takes two arguments. The first is the “boss” of the current subtree (let’s call him Bob). The second argument is a list of the results
+for each subtree under Bob. Each result is a pair of GuestLists: the
+first GuestList in the pair is the best possible guest list with the boss
+of that subtree; the second is the best possible guest list without the
+boss of that subtree. nextLevel should then compute the overall best
+guest list that includes Bob, and the overall best guest list that doesn’t
+include Bob.
+-}
+nextLevel :: Employee -> [(GuestList, GuestList)] -> (GuestList, GuestList)
+nextLevel boss bestRes = (withBoss, withoutBoss)
+  where
+    withBoss :: GuestList
+    withBoss    = glCons boss (maximumGL $ map snd bestRes)
+
+    withoutBoss :: GuestList
+    withoutBoss = maximumGL $ map fst bestRes
```

Comment:
If I understand the algorithm right, this implementation seems to be incorrect. If `boss` is not going to the party, then you should `mconcat` all the guest lists from the subtrees. On the other hand if boss is going to the party, then you can `moconcat` all the `withoutBoss` values from the subtree and `mappend` the boss onto it.

Reply:
@jhrcek Yep, you are right. I missed out on this sentence at the top: _It is a well-known fact that anyone whose immediate boss is also at the party will not have any fun at all_. I have updated my attempt accordingly.